dace/sdfg/state.py

```diff
@@ -1548,42 +1576,123 @@ def add_memlet_path(self,
-                              dtype,
-                              storage,
-                              transient,
-                              strides,
-                              offset,
-                              lifetime,
-                              debuginfo,
-                              find_new_name=find_new_name,
-                              total_size=total_size,
-                              alignment=alignment)
-        return self.add_access(name, debuginfo)
+    def remove_memlet_path(self,
+                           edge: MultiConnectorEdge,
+                           remove_orphans: bool = True) -> None:
+        """ Removes all memlets and associated connectors along a path formed
+            by a given edge. Undefined behavior if the path is ambiguous.
+            Orphaned entry and exit nodes will be connected with empty edges to
+            maintain connectivity of the graph.
+
+            :param edge: An edge that is part of the path that should be
+                         removed, which will be passed to `memlet_path` to
+                         determine the edges to be removed.
+            :param remove_orphans: Remove orphaned data nodes from the graph if
+                                   they become orphans from removing this memlet
+                                   path.
+        """
+
+        path = self.memlet_path(edge)
+
+        is_read = isinstance(path[0], nd.AccessNode)
+        if is_read:
+            # Traverse from connector to access node, so we can check if it's
+            # safe to delete edges going out of a scope
+            path = reversed(path)
+
+        for edge in path:
+
+            self.remove_edge(edge)
+
+            # Check if there are any other edges exiting the source node that
+            # use the same connector
+            other_outgoing = False
+            for e in self.out_edges(edge.src):
+                other_outgoing = True
+                if e.src_conn is not None and e.src_conn == edge.src_conn:
+                    other_outgoing_same = True
+                    break
+            else:
+                other_outgoing_same = False
+                edge.src.remove_out_connector(edge.src_conn)
+
+            # Check if there are any other edges entering the destination node
+            # that use the same connector
+            other_incoming = False
+            for e in self.in_edges(edge.dst):
+                other_incoming = True
+                if e.dst_conn is not None and e.dst_conn == edge.dst_conn:
+                    other_incoming_same = True
+                    break
+            else:
+                other_incoming_same = False
+                edge.dst.remove_in_connector(edge.dst_conn)
+
+            if isinstance(edge.src, nd.EntryNode):
+                # If removing this edge orphans the entry node, replace the
+                # edge with an empty edge
+                if not other_outgoing:
```

Comment:
the condition here should be `if not other_incoming:`.
Think of the following case:
```
       (A) (B)
      /  Map  \
       |     |
       v     v
     /M1 \  / M2\
```

Let's say you remove the right edge (B->Map->M2), then M2 will be disconnected because Map has outgoing edges.

Reply:
Shouldn't it actually be `if len(self.edges_between(src, dst)) == 0` in both cases? The important thing is to not break any scopes.

Reply:
@tbennun What do you think of 37607217?

Reply:
Much better. Captures exactly what you want (disconnection)